### PR TITLE
Improve cli.

### PR DESCRIPTION
**Pull request by olafurpg**
- tab completion for zsh
jun-19-2017 09-30-49
- tab completion for bash
jun-19-2017 09-52-01 bash

- --files did not filter out .scala files.
- the custom error message for semantic rewrites did not appear
  for custom rewrites.
- unified --files and --include
- update --help on --classpath (was outdated)

### CODE REVIEW

**Comment by olafurpg** [comment]
Will need to continue this next week Monday.

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala [hunk 1]
```diff
@@ -29,6 +53,93 @@ object Cli {
...
+        return 0
+        ;;
+    esac
+    if [[ $${cur} == -* ]] ; then
+        COMPREPLY=(   $$(compgen -W "$${opts}" -- $${cur}) )
+        return 0
+    fi
+}
+complete -F _scalafix scalafix
+""".stripMargin
```
Comment (allanrenucci) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: There is no margin to strip here

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala [hunk 2]
```diff
@@ -29,6 +53,93 @@ object Cli {
...
+scalafix_opts=(
+  $zshArgs
+)
+
+case $$words[$$CURRENT] in
+      *) _arguments $$scalafix_opts "*::filename:_files";;
+esac
+
+return 0
+""".stripMargin
```
Comment (allanrenucci) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: There is no margin to strip here

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala (hunk 1 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: Thanks! Removed.

File: scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala [hunk 3]
```diff
@@ -203,18 +249,37 @@ object CliRunner {
...
         }
-        Ok(Classpath(paths))
-    }
+        if (cp.shallow.nonEmpty) Some(Ok(cp))
+        else {
+          val msg =
+            """Unable to automatically detect .semanticdb files to run semantic rewrites. Possible workarounds:
+              |- re-compile sources with the scalahost compiler plugin enabled.
+              |- pass in the --syntactic flag to run only syntactic rewrites.
+              |- explicitly pass in --classuath and --sourceroot to run semantic rewrites.""".stripMargin
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: Typo

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala [hunk 4]
```diff
@@ -5,19 +5,43 @@ import scala.meta.Database
...
+    s"""|Available rewrites: ${ScalafixRewrites.allNames.mkString(", ")}
+        |
+        |NOTE. The command line tool is mostly intended to be invoked programmatically
+        |from build-tool integrations such as sbt-scalafix. The necessary fixture to run
+        |semantic rewrites is tricky to setup manually.
+        |
+        |Scalafix chooses which files to fix according to the following rules:
+        |- when --syntactic is passed, then Scalafix looks for .scala files in the provided files/directories.
+        |- by default, looks for .semanticdb files and matches them back to the original files.
+        |  - if --classpath and --sourcepath are provided, then those are used to find .semanticdb files.
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: `--sourceroot`

**Comment by xeno-by** [comment]
`scalafix -f --help` says "Running scalafix on 0 files" and exits without printing any help. Not sure whether this is intentional.

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala [hunk 5]
```diff
@@ -5,19 +5,43 @@ import scala.meta.Database
...
+        |  - otherwise, Scalafix will automatically look for META-INF/semanticdb directories from the
+        |    current working directory.
+        |
+        |Examples (semantic):
+        |  $$ scalafix # automatically finds .semanticdb files and runs rewrite configured in .scalafix.conf.
+        |  $$ scalafix <directory> # same as above except only run on files in <directory>
+        |  $$ scalafix --rewrites RemoveUnusedImports # same as above but run RemoveUnusedImports.
+        |  $$ scalafix --classpath <foo.jar:target/classes> --sourceroot <directory> # customize semanticdb.
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: Does `--sourceroot` have a default value?

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala [hunk 6]
```diff
@@ -5,19 +5,43 @@ import scala.meta.Database
...
+        |
+        |Examples (syntactic):
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: Can `--syntactic` be autodetected from the list of rewrites that are requested by the user?

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala [hunk 7]
```diff
@@ -5,19 +5,43 @@ import scala.meta.Database
...
+        |  $$ scalafix --syntactic --rewrites=ProcedureSyntax Code.scala # write fixed file in-place
+        |  $$ scalafix --syntactic --rewrites=ProcedureSyntax --stdout Code.scala # print fixed file to stdout
+        |  $$ scalafix --syntactic --rewrites=ExplicitReturnTypes --auto-mirror # automatically configure Semantic API.
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: Why is `ExplicitReturnTypes` considered to be syntactic?

File: scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala [hunk 8]
```diff
@@ -44,85 +50,87 @@ case class ScalafixOptions(
...
     sourceroot: Option[String] = None,
     @HelpMessage(
-      """java.io.File.pathSeparator separated list of jar files or directories
-        |        containing classfiles and `semanticdb` files. The `semanticdb`
-        |        files are emitted by the scalahost-nsc compiler plugin and
-        |        are necessary for the semantic API to function. The
-        |        classfiles + jar files are necessary forruntime compilation
-        |        of quasiquotes when extracting symbols (that is,
-        |        `q"scala.Predef".symbol`)""".stripMargin
+      "java.io.File.pathSeparator separated list of directories containing " +
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala]: Just directories? Or jars are also supported?

File: scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala [hunk 9]
```diff
@@ -91,24 +97,63 @@ sealed abstract case class CliRunner(
...
+          val fileToWrite = scala.io.Source.fromFile(outFile)
+          val originalContents = input.chars.toIterator
+          try {
+            @tailrec
+            def isEqual(a: Iterator[Char], b: Iterator[Char]): Boolean =
+              if (a.hasNext && b.hasNext) a.next() == b.next() && isEqual(a, b)
+              else if (a.hasNext || b.hasNext) false
+              else true
+            isEqual(fileToWrite, originalContents)
+          } finally fileToWrite.close()
```
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: Why not read the entire file into memory and compare strings? That would probably be simpler.

Comment (allanrenucci) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: Or
```scala
try fileToWrite.sameElements(input.chars)
finally fileToWrite.close()
```

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala (hunk 4 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: Fixed.

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala (hunk 6 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: That is an interesting idea. It's simple for provided rewrites but more tricky for classloaded rewrites. I think it's best to leave that for another PR, https://github.com/scalacenter/scalafix/issues/221

File: scalafix-cli/src/main/scala/scalafix/cli/Cli.scala (hunk 7 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/Cli.scala]: outdated line, removed. --auto-mirror doesn't even exist anymore.

File: scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala [hunk 10]
```diff
@@ -50,8 +54,8 @@ sealed abstract case class CliRunner(
     config: ScalafixConfig,
     database: Option[Database],
     rewrite: Rewrite,
-    explicitPaths: Seq[Input],
-    inputs: Seq[Input],
+    explicitPaths: scala.Seq[Input],
```
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: List for meta trees will be really nice @xeno-by

File: scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala (hunk 9 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: Nice @allanrenucci  I didn't know that was possible, thanks!

Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: @xeno-by It's unnecessary to load everything into memory when we can do it char by char.

File: scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala (hunk 3 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: Fixed.

File: scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala (hunk 8 above)
Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala]: Jars are also supported, updated.

Comment (olafurpg) [inline scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala]: Although, I have no idea what will happen if a jar is passed in. I suspect scalafix will try to fix those files too :O Do we have any tests in scalameta/scalameta for reading/writing .semanticdb files from jars?

File: scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala (hunk 10 above)
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala]: List or Seq?

File: scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala (hunk 8 above)
Comment (xeno-by) [inline scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala]: Probably not, but last time I checked they worked :)

**Comment by olafurpg** [comment]
There are still two remaining improvements that I will leave for other PRs
- find explicit paths first, then match against .semanticdb files
- automatically detect --syntactic if passed rewrites are syntactic only
